lspserver/include/lspserver/LSPBinder.h

```diff
@@ -12,8 +12,8 @@ namespace lspserver {
 template <typename T>
 llvm::Expected<T> parseParam(const llvm::json::Value &Raw,
                              llvm::StringRef PayloadName,
-                             llvm::StringRef PayloadKind) {
-  T Result;
+                             llvm::StringRef PayloadKind, T Base = T()) {
+  T Result = Base;
```

Comment:
I would argue that it's reasonable to have an optional argument that defines the defaults for parsing. It is an optional parameter so it shouldn't change the behaviour of the function (although maybe move semantics could be used to make it more efficient).

If I can't change this then I will either have to copy the implementation of this function into Controller.cpp (breaks DRY) or make a more complicated system of defaulting that acts on the parsed `configuration::TopLevel` structs (more error-prone and less efficient). Either of these options would make the code less maintainable. What would you suggest?

Reply:
> structs (more error-prone and less efficient).

How about use a dedicated function like `updateConfig` and do the work there? I think introduce a new default argument in the basic library is more error prone for large C++ project (since it might implicitly change many things)

Reply:
I have refactored `parseParam` again so that it no longer takes a default argument and behaves the same way as before (uninitializing `Result` by default or using the default constructor if this is not possible, as well as not compiling if neither is possible). However, it now shares code with `parseParamWithDefault`, which has an argument that determines some default JSON that should be used. I think that this is a reasonable function to add to a general-purpose LSP library.